File: algorithms/algorithms-cplusplus/leetcode/Question_0590_N_ary_Tree_Postorder_Traversal.cc
#include <vector>
#include <stack>

using namespace std;

// Definition for a Node.
class Node {
 public:
  int val;
  vector<Node*> children;

  Node() {}

  Node(int _val) {
      val = _val;
  }

  Node(int _val, vector<Node*> _children) {
      val = _val;
      children = _children;
  }
};

class Solution {
 public:
  vector<int> postorder(Node* root) {
    vector<int> res;
    help(root, res);
    return res;
  }

  void help(Node* root, vector<int>& res) {
    if (!root) {
      return;
    }
    for (Node* p : root->children) {
      help(p, res);
    }
    res.push_back(root->val);
  }

  vector<int> postorder2(Node* root) {
    vector<int> res;
    if (!root) {
      return res;
    }
    stack<Node*> s;
    s.push(root);
    while (!s.empty()) {
      root = s.top();
      s.pop();
      res.push_back(root->val);
      for (int i = 0; i < root->children.size(); i++) {
        s.push(root->children[i]);
      }
    }
    return vector<int>(res.rbegin(), res.rend());
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0590_N_ary_Tree_Postorder_Traversal.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

public class Question_0590_N_ary_Tree_Postorder_Traversal {
    class Node {
        public int val;
        public List<Node> children;

        public Node() {}

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }

    public List<Integer> postorder(Node root) {
        List<Integer> res = new ArrayList<>();
        help(root, res);
        return res;
    }

    private void help(Node root, List<Integer> res) {
        if (root == null) {
            return;
        }
        for (Node p : root.children) {
            help(p, res);
        }
        res.add(root.val);
    }

    public List<Integer> preorder2(Node root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Stack<Node> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            root = stack.pop();
            res.add(root.val);
            for (int i = 0; i < root.children.size(); i++) {
                stack.push(root.children.get(i));
            }
        }
        Collections.reverse(res);
        return res;
    }
}
